File: 백준/Gold/1043. 거짓말/거짓말.js
const fs = require("fs");
const filePath = process.platform === "linux" ? "/dev/stdin" : "./input.txt";
let input = fs.readFileSync(filePath).toString().trim().split("\n");

const [N, M] = input.shift().split(" ").map(Number);
const truth = input.shift().split(" ").map(Number).slice(1);
const parties = input.map((el) => el.split(" ").map(Number).slice(1));
const graph = Array.from({ length: N + 1 }, () => new Set());

const helper = (party) => {
  for (let i = 0; i < party.length - 1; i++) {
    for (let j = i + 1; j < party.length; j++) {
      graph[party[i]].add(party[j]);
      graph[party[j]].add(party[i]);
    }
  }
};

for (let party of parties) {
  if (party.length < 2) continue;
  helper(party);
}

const dfs = (visited, cur) => {
  let result = false;
  if (truth.includes(cur)) return true;
  visited[cur] = true;

  for (let next of graph[cur]) {
    if (visited[next]) continue;
    result = dfs(visited, next);
    if (result === true) return result;
  }

  return result;
};

let count = 0;
for (let party of parties) {
  let lie = true;
  for (let person of party) {
    const visited = Array(N + 1).fill(false);
    if (dfs(visited, person)) {
      lie = false;
      break;
    }
  }
  if (lie) {
    count++;
  }
}

console.log(count);

File: 백준/Gold/1043. 거짓말/거짓말.py
import sys
from itertools import combinations

input = sys.stdin.readline

n, m = map(int, input().split())
liars = list(map(int, input().split()[1:]))

parties = [list(map(int, input().split()[1:])) \
  for _ in range(m)]

graph = [set() for _ in range(n+1)]
for party in parties:
  if len(party) < 2:
    continue

  all_case = combinations(party, 2)
  for x, y in all_case:
    graph[x].add(y)
    graph[y].add(x)

def dfs(visited, liars, x): # True -> liar를 만남
  result = False
  if x in liars:
    return True
  
  visited[x] = True
  for next_node in graph[x]:
    if visited[next_node]:
      continue
    result = dfs(visited, liars, next_node)
    if result == True:
      return result
  
  return result
  
count = 0
for party in parties:
  for player in party:
    visited = [False] * (n+1)
    if dfs(visited, liars, player):
      break
  else:
    count += 1

print(count)